### PR TITLE
Add functionality to explicitely specify a span parent

### LINKED ISSUE #318: Add functionality to explicitly parent a span from a parent span or parent context
According to the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#span-creation), the `Tracer` must provide the possibility to provide a parent span or a parent span context on span creation.

### PR DESCRIPTION
**Pull request by pyohannes**
This PR extends the `StartSpanOptions` struct by a `parent` SpanContext. If this is given, it will be used as a parent for a given span.

While I was at it, I cleaned up some things in the `SpanContext` implementation.

Closes #318.

### CODE REVIEW

File: api/include/opentelemetry/trace/span_context.h [hunk 1]
```diff
@@ -33,10 +33,6 @@ namespace trace_api = opentelemetry::trace;
 class SpanContext final
 {
 public:
-  // An invalid SpanContext.
-  SpanContext() noexcept
```
Comment (pyohannes) [inline api/include/opentelemetry/trace/span_context.h:37]: Let's use `SpanContext::GetInvalid()` as the default way to create an invalid `SpanContext`. Having both `SpanContext()` and `SpanContext::GetInvalid()` returning an invalid `SpanContext` is somewhat confusing.

File: api/include/opentelemetry/trace/span_context.h [hunk 2]
```diff
@@ -110,10 +84,10 @@ class SpanContext final
...
 
 private:
-  const trace_api::TraceId trace_id_;
-  const trace_api::SpanId span_id_;
-  const trace_api::TraceFlags trace_flags_;
-  const bool remote_parent_ = false;
+  trace_api::TraceId trace_id_;
+  trace_api::SpanId span_id_;
+  trace_api::TraceFlags trace_flags_;
+  bool remote_parent_ = false;
```
Comment (pyohannes) [inline api/include/opentelemetry/trace/span_context.h:90]: Removing the `const` enables us to use default (and hopefully optimized) copy constructors and assignment operators.

File: api/include/opentelemetry/trace/span_context.h (hunk 1 above)
Comment (ThomsonTan) [reply api/include/opentelemetry/trace/span_context.h:37]: Explicit delete as `SpanContext() = delete;`?

File: api/include/opentelemetry/trace/span.h [hunk 3]
```diff
@@ -43,8 +43,13 @@ struct StartSpanOptions
   core::SystemTimestamp start_system_time;
   core::SteadyTimestamp start_steady_time;
 
+  // Explicitely set the parent of a Span.
+  //
+  // This defaults to an invalid span context. In this case, the Span is
+  // automatically parented to the currently active span.
+  SpanContext parent = SpanContext::GetInvalid();
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/span.h:50]: Defaults to invalid value may look a little weird, would `GetEmpty()` a better name?

Comment (pyohannes) [reply api/include/opentelemetry/trace/span.h:50]: I was thinking about that, but then decided against introducing an _empty_ span context. In the current terminology, a span context can either be _valid_ or _invalid_ (it has an `IsValid` method). So `SpanContext::GetInvalid()::IsValid() == true` I think is more idiomatic than `SpanContext::GetEmpty()::IsValid() == true`. And here, an invalid span context is used like a null pointer (indicating that no parent span context is set).

But I don't have a very strong opinion on that, I'd be willing to go with `GetEmpty()` if folks feel that's clearer.

File: api/include/opentelemetry/trace/span_context.h (hunk 1 above)
Comment (pyohannes) [reply api/include/opentelemetry/trace/span_context.h:37]: Good call, I made it explicit.

File: api/include/opentelemetry/trace/span.h (hunk 3 above)
Comment (lalitb) [reply api/include/opentelemetry/trace/span.h:50]: I too don't have strong opinion on this, although having two different methods `GetEmpty()` and `GetInvalid()` returning invalid span context is more confusing to me.

Comment (lalitb) [inline api/include/opentelemetry/trace/span.h:50]: Just wondering if it would be more optimized to have  `SpanContext::GetInvalid()` method returning a static instance of invalid context ( using c++11 local atomic static ). Can be part of this or separate PR.

File: api/include/opentelemetry/trace/span_context.h [hunk 4]
```diff
@@ -71,31 +67,9 @@ class SpanContext final
...
-  //
-  //  SpanContext &operator=(SpanContext &&ctx)
-  //  {
-  //    SpanContext *spn_ctx =
-  //        new SpanContext(ctx.trace_id(), ctx.span_id(), ctx.trace_flags(),
-  //        ctx.HasRemoteParent());
-  //    this = spn_ctx;
-  //    return *this;
-  //  };
+  SpanContext &operator=(const SpanContext &ctx) = default;
```
[the remark this answers is not in the document]
Comment (pyohannes) [reply api/include/opentelemetry/trace/span_context.h:72]: Having an assignment operator makes it easier and more efficient to specify a parent context in the `StartSpanOptions` struct:

```c++
trace_api::StartSpanOptions options;
options.parent   = span_first->GetContext();
```

Without an assignment operator, we'd need to resort to a pointer (a heap allocated `SpanContext` object), which is more inefficient (as it requires a heap allocation/deallocation).

The `SpanContext` is immutable in the sense, that it has no setters (span id, trace id and flags cannot be changed independently of each other), and the `SpanContext` assigned to a `Span` cannot change either.

File: api/include/opentelemetry/trace/span.h (hunk 3 above)
Comment (pyohannes) [reply api/include/opentelemetry/trace/span.h:50]: That would be more efficient if we'd return a pointer, but as we return by value here this involves a copy anyway, and we wouldn't gain much by having a static singleton.

**Review by lalitb — APPROVED** [review]

**Review by ThomsonTan — APPROVED** [review]
